pkg/apis/machine/v1beta1/machine_webhook.go

```diff
@@ -534,3 +567,202 @@ func validateAzure(h *validatorHandler, m *Machine) (bool, utilerrors.Aggregate)
 	}
 	return true, nil
 }
+
+func defaultGCP(h *defaulterHandler, m *Machine) (bool, utilerrors.Aggregate) {
+	klog.V(3).Infof("Defaulting GCP providerSpec")
+
+	var errs []error
+	providerSpec := new(gcp.GCPMachineProviderSpec)
+	if err := unmarshalInto(m, providerSpec); err != nil {
+		errs = append(errs, err)
+		return false, utilerrors.NewAggregate(errs)
+	}
+
+	if providerSpec.MachineType == "" {
+		providerSpec.MachineType = defaultGCPMachineType
+	}
+
+	if len(providerSpec.NetworkInterfaces) == 0 {
+		providerSpec.NetworkInterfaces = append(providerSpec.NetworkInterfaces, &gcp.GCPNetworkInterface{
+			Network:    defaultGCPNetwork(h.clusterID),
+			Subnetwork: defaultGCPSubnetwork(h.clusterID),
+		})
+	}
+
+	providerSpec.Disks = defaultGCPDisks(providerSpec.Disks, h.clusterID)
+
+	if len(providerSpec.Tags) == 0 {
+		providerSpec.Tags = defaultGCPTags(h.clusterID)
+	}
+
+	if providerSpec.UserDataSecret == nil {
+		providerSpec.UserDataSecret = &corev1.LocalObjectReference{Name: defaultUserDataSecret}
+	}
+
+	if providerSpec.CredentialsSecret == nil {
+		providerSpec.CredentialsSecret = &corev1.LocalObjectReference{Name: defaultGCPCredentialsSecret}
+	}
+
+	if len(providerSpec.ServiceAccounts) == 0 {
+		providerSpec.ServiceAccounts = defaultGCPServiceAccounts(h.clusterID, providerSpec.ProjectID)
+	}
+
+	rawBytes, err := json.Marshal(providerSpec)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	if len(errs) > 0 {
+		return false, utilerrors.NewAggregate(errs)
+	}
+
+	m.Spec.ProviderSpec.Value = &runtime.RawExtension{Raw: rawBytes}
+	return true, nil
+}
+
+func defaultGCPDisks(disks []*gcp.GCPDisk, clusterID string) []*gcp.GCPDisk {
+	if len(disks) == 0 {
+		return []*gcp.GCPDisk{
+			{
+				AutoDelete: true,
+				Boot:       true,
+				SizeGb:     defaultGCPDiskSizeGb,
+				Type:       defaultGCPDiskType,
+				Image:      defaultGCPDiskImage(clusterID),
+			},
+		}
+	}
+
+	for _, disk := range disks {
+		if disk.Type == "" {
+			disk.Type = defaultGCPDiskType
+		}
+
+		if disk.Image == "" {
+			disk.Image = defaultGCPDiskImage(clusterID)
+		}
+	}
+
+	return disks
+}
+
+func validateGCP(h *validatorHandler, m *Machine) (bool, utilerrors.Aggregate) {
+	klog.V(3).Infof("Validating GCP providerSpec")
+
+	var errs []error
+	providerSpec := new(gcp.GCPMachineProviderSpec)
+	if err := unmarshalInto(m, providerSpec); err != nil {
+		errs = append(errs, err)
+		return false, utilerrors.NewAggregate(errs)
+	}
+
+	if providerSpec.Region == "" {
+		errs = append(errs, field.Required(field.NewPath("providerSpec", "region"), "region is required"))
+	}
+
+	if !strings.HasPrefix(providerSpec.Zone, providerSpec.Region) {
+		errs = append(errs, field.Invalid(field.NewPath("providerSpec", "zone"), providerSpec.Zone, fmt.Sprintf("zone not in configured region (%s)", providerSpec.Region)))
+	}
+
+	if providerSpec.ProjectID == "" {
```

Comment:
the projectID is not required. It might come from serviceAccount given by the creds secret -> https://github.com/openshift/cluster-api-provider-gcp/blob/e1f37971a88a65c7bf134983ee8a8414d89866a4/pkg/cloud/gcp/actuators/machine/machine_scope.go#L63-L69

Reply:
it comes from `GCP *gcp.Platform`

Reply:
Ack, removed that requirement, allowed it to default service accounts still if the project ID is given, do you think that's worth it or get rid?

Reply:
this is available in `infrastructure.Status.PlatformStatus.GCP`, how about we use it to default the serviceAccountName

Reply:
Ahh right, I see what you mean, should we default the ProjectID as well or is that not really worth it?